shopinvader/utils.py

```diff
@@ -0,0 +1,82 @@
+# Copyright 2021 Camptocamp (http://www.camptocamp.com).
+# @author Simone Orsi <simahawk@gmail.com>
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl).
+from contextlib import contextmanager
+
+from odoo.addons.base_rest.controllers.main import _PseudoCollection
+from odoo.addons.component.core import WorkContext
+
+
+def get_partner_work_context(shopinvader_partner):
+    """Retrieve service work context for given shopinvader.partner
+    """
+    ctx = {}
+    ctx["invader_partner"] = shopinvader_partner
+    ctx["invader_partner_user"] = shopinvader_partner
+    partner_user = shopinvader_partner.record_id
+    ctx["partner_user"] = partner_user
+    # The partner user for the main account or for sale order may differ.
+    partner_shop = partner_user.get_shop_partner(
+        shopinvader_partner.backend_id
+    )
+    ctx["partner"] = partner_shop
+    if partner_shop != partner_user:
+        # Invader partner must rappresent the same partner as the shop
+        invader_partner_shop = partner_shop._get_invader_partner(
+            shopinvader_partner.backend_id
+        )
+        if invader_partner_shop:
+            ctx["invader_partner"] = invader_partner_shop
+    return ctx
+
+
+def load_partner_work_ctx(service, invader_partner, force=False):
+    """Update work context for given service loading given invader partner ctx."""
+    params = get_partner_work_context(invader_partner)
+    update_work_ctx(service, params, force=force)
+
+
+def reset_partner_work_ctx(service):
+    """Update work context flushing all partner keys."""
+    defaults = {}
+    partner_work_context_defaults(
+        service.env, service.shopinvader_backend, defaults
+    )
+    update_work_ctx(service, defaults, force=True)
+
+
+def update_work_ctx(service, params, force=False):
+    """Update work context for given service."""
+    for k, v in params.items():
+        if force or not getattr(service.work, k, None):
+            setattr(service.work, k, v)
+
+
+def partner_work_context_defaults(env, backend, params):
+    """Inject defaults as these keys are mandatory for work ctx."""
+    if params.get("partner") and not params.get("partner_user"):
```

Comment:
```suggestion
    if params.get("partner") and params.get("partner_user") is None:
```
I don't know if it can hurt, but…

Reply:
here I don't care. No matter which value gives "null" (empty recordset, false or none) I want to be sure it gets its own default.